druid-shell/src/alert.rs

```diff
@@ -0,0 +1,451 @@
+/// In window scope this means the user can't interact with the parent window
+/// and in application scope this means the user can't interact with any other window.
+///
+/// # Multiple alerts
+///
+/// There is no guaranteed behavior when multiple alert dialogs are requested with
+/// overlapping modality scope. Depending on platform specifics, all the alerts
+/// might be visible or just some of them. Regardless of visibility, only one of
+/// the alert dialogs will be interactable. When that alert dialog is closed,
+/// another one will become interactable.
+///
+/// If the modality scopes don't overlap, i.e. when multiple alerts are originating
+/// from different windows and they are all window scoped, all the alerts can be
+/// interacted with and they won't block eachother.
+#[derive(Debug, Clone)]
+pub struct AlertOptions {
+    /// Whether the alert is app-modal.
+    pub(crate) app_modal: bool,
+    /// The context of the alert.
+    pub(crate) context: String,
+    /// The primary message of the alert.
+    pub(crate) message: String,
+    /// The supplemental text of the alert.
+    pub(crate) description: String,
+    /// The icon of the alert.
+    pub(crate) icon: Option<AlertIcon>,
+    /// The primary button of the alert.
+    pub(crate) primary: AlertButton,
+    /// The cancel button of the alert.
+    pub(crate) cancel: Option<AlertButton>,
+    /// The alternative buttons of the alert.
+    pub(crate) alternatives: Vec<AlertButton>,
+}
+
+/// Alert dialog icon.
+///
+/// # macOS
+///
+/// On macOS the information and warning icons are the same.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum AlertIcon {
+    /// The alert dialog is providing important time sensitive information.
+    Information,
+    /// The alert dialog is warning about a potential future problem.
+    Warning,
+    /// The alert dialog is informing of a critical problem that has already happened.
+    Error,
+}
+
+/// A specific button for the alert dialog.
+///
+/// If the alert is just reporting information and it only needs a single button,  
+/// then you should use a primary button with the label **OK** or a translation of it.
+///
+/// The cancel button should always be labeled **Cancel** or a translation of it.
+///
+/// In all other cases you should give a short but descriptive label that includes a verb.
+/// The label should let the user know what will happen when they click the button.
+/// Assume that the user didn't actually read the alert message. This assumption ends up
+/// true more often than not, because alerts by design appear unexpectedly while the user
+/// is trying to accomplish some other task.
+///
+/// Labels like **Send message**, **View all**, **Reconnect**, **Remind me tomorrow**, and
+/// **Delete file** make it much more clear what will happen compared to a generic **OK**,
+/// **Confirm**, or **Yes**.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AlertButton {
```

Comment:
According to the  [style guide](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#avoid-redundant-prefixes-[rfc-356]) we should be calling these just `Button` or `Options` and then use them like `alert::Button` or `alert::Options`.

This will of course mean that the naming will collide with `Button` if you were to import everything into the same scope, but I don't think typing `alert::Button` is worse than `AlertButton`.

This would also make sens if we would not reexport everything in the crate root, because then you would have to use `alert::AlertButton` which is pretty redundant.

So if we want to continue reexporting everything, `AlertButton` makes sense, otherwise we should remove the prefix.

Reply:
I think for this PR they should remain like `AlertButton` as we have `WidgetId`, `FileInfo`, `FlexParams` etc.

However this is worth having a discussion over whether we should rename everything and also stop re-exporting everything to the root namespace and organize things more nicely into modules.

We should open a separate issue for this as a general problem+discussion. Feel free to get that started.

Reply:
I've opened an issue for it (#1051).